Base/usr/share/man/man5/Shell.md

```diff
@@ -220,6 +220,35 @@ $ { for * { te $it } }&
 $ disown
 ```
 
+##### Functions
+A function is a user-defined entity that can be used as a simple command to execute a compound command, optionally with some parameters.
+Such a function is defined via the syntax below:
+
+```sh
+function_name(explicitly_named_arguments...) { compound_command }
+```
+
+The function is named `function_name`, and has some explicitly named arguments `explicitly_named_arguments...`, which *must* be supplied by the caller, failure to do so will cause the command to exit with status 1.
+
+The compound command shall be executed whenever the simple command `function_name` is executed.
+This execution shall be performed in a subshell, in a new local frame.
```

Comment:
Wait, why in a subshell?

Reply:
is it supposed to be run in the current shell process itself?
how would job control work with that?

Reply:
Uh, I've always thought shells run functions in-process, just like other control flow, or like other languages

Not sure about job control; do you allow job control on builtins (e.g. `cd foo/ &`?)

Reply:
> Job control on builtins

Hmmm...I don't think so, no.

Reply:
Is there any good reason to make functions run in-process? the only thing I can think of is to make modifying the shell state possible...
But always running it in a subshell has a nice uniformity; piping two functions together wouldn't be possible if we make them run in-process

Reply:
If the user tries to use I/O redirections (pipes) or job control, *then* run it in a subshell. Or even require the user to do it explicitly: `{ fn 1 2 } | head`, but I'm not sure that's a good idea.

What do other shells do?

Reply:
It seems zsh does as you're suggesting, if it's run as part of a pipeline or something that would require job control, it's run in a subshell.
otherwise it's run in the current process.

Reply:
An alternative way would be to simply make a syntactical distinction between the two, one would be always run in the current process, and the other always in a subprocess

Reply:
> If the user tries to use I/O redirections (pipes) or job control, then run it in a subshell.

Ah heck it, I'll just do this. seems to be the most sane option (according to other shells) of the bunch